Sources/TuistGenerator/Generator/Hook/Carthage.swift

```diff
@@ -0,0 +1,38 @@
+import Basic
+import TuistCore
+
+public struct CarthageHook: GenerateHook {
+    
+    public let owner: Generating
+    
+    public init(generator: Generating) {
+        self.owner = generator
+    }
+
+    private let carthageInteractor: CarthageInteracting = CarthageInteractor()
+    
+    public func pre(path: AbsolutePath) throws {
+        try carthageInteractor.checkout()
+    }
+    
+    public func post(path: AbsolutePath) throws {
+        
+        guard let graph = Memoized.graph[path] else {
+            return
+        }
+        
+        let hasCarthageDependencies = graph.carthageDependencies.isEmpty == false
+        let usingCarthageProjects = CLI.arguments.carthage.projects == false
```

Comment:
I introduced the CLI arguments to reduce the number of injections throughout - I treat the CLI arguments as `static` data even if they are not.

I do agree with you that we need to find a way to inject them into TusitGenerator so that it stays independent of the arguments. CarthageHook would be good to pull out but as mentioned in https://github.com/tuist/tuist/pull/579#discussion_r336890249 there's a dependency on the Graph class to figure out the different platforms to build for.

I was also going to convert CocoapodsInteractor to use the same mechanism, but that also uses the graph. Currently only the Graph is visible to TuistGenerator and I wasn't overly keen to expose that.